### ISSUE TITLE
detshuffle epoch count the same across epochs

### ISSUE THREAD

**Issue by rwightman** [webdataset/webdataset#172]
I'm using a pipeline as per #169, the core of it isn't disimilar to the OpenImages notebook (https://github.com/webdataset/webdataset/blob/ccfe88086cdb21a0dc23a6454ce3e3723b6b8033/notebooks/openimages.ipynb) except I have extra shuffle and worker/node splitters, etc.

I was noticing some worsened accuracy numbers after transitioning from older wds 0.1 code... and I think I've pinned it down to improper shuffling, the shards are not being shuffled across nodes and workers as expected. The epoch coutn in detshuffle is never incrementing, it always starts from -1 and is incremented to 0 in each epoch. 

I haven't pinned down the root cause yet, in a debugger, on every epoch the invoke calls run on detshuffle instance and self.epoch is -1. Yet, I haven't seen the detshuffle init method called after the first epoch. First though is that the data structure gets copied from the original instance each epoch. It's a bit of a chain to follow so haven't pinpointed that

@tmbdev any ideas? this seems significant unless I' competely misunderstanding the correct use here, the previous method where an explict epoch set via env var was working, this is a cleaner API but seems to have a gotcha.

**Comment by tmbdev** [comment]
Yes, unfortunately, there are lots of subtle interactions in the way the Dataloader handles worker processes and internal state right now (torchdata will clean this up). I'll have to add something to the documentation.

The recommended way is to just use resampling if you can, since that has no configuration and will always do something reasonable.

**Comment by rwightman** [comment]
@tmbdev ugly as it is, it might be worth continuing to support use of `WDS_EPOCH` in the run() of detshuffle, it would work in all cases... it could be used only if set, otherwise fallback to seeding with the  `self.epoch counter` (with some warning in documentation to describe the failure case for `persistent_workers=False`
